pkg/controller/configuration/controller.go

```diff
@@ -0,0 +1,167 @@
+	"github.com/onosproject/onos-config/pkg/utils/jsonvalues"
+	"github.com/onosproject/onos-lib-go/pkg/logging"
+)
+
+var log = logging.GetLogger("controller", "configuration")
+
+const (
+	defaultTimeout = 30 * time.Second
+)
+
+// NewController returns a configuration controller
+func NewController(topo topo.Store, conns gnmi.ConnManager, configurations configuration.Store) *controller.Controller {
+	c := controller.NewController("configuration")
+
+	c.Watch(&Watcher{
+		configurations: configurations,
+	})
+
+	c.Watch(&TopoWatcher{
+		topo: topo,
+	})
+
+	c.Reconcile(&Reconciler{
+		conns:          conns,
+		topo:           topo,
+		configurations: configurations,
+	})
+
+	return c
+}
+
+// Reconciler reconciles configurations
+type Reconciler struct {
+	conns          gnmi.ConnManager
+	topo           topo.Store
+	configurations configuration.Store
+}
+
+// Reconcile reconciles configurations
+func (r *Reconciler) Reconcile(id controller.ID) (controller.Result, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTimeout)
+	defer cancel()
+
+	configurationID := id.Value.(configapi.ConfigurationID)
+	config, err := r.configurations.Get(ctx, configurationID)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			log.Warnf("Failed to reconcile configuration %s, %s", configurationID, err)
+			return controller.Result{}, err
+		}
+		log.Debugf("Configuration %s not found", configurationID)
+		return controller.Result{}, nil
+	}
+	log.Infof("Reconciling configuration %v", config)
+	if config.Status.State != configapi.ConfigurationState_CONFIGURATION_PENDING {
+		log.Debugf("Skipping Reconciliation of configuration %s, configuration state is: %s", configurationID, config.Status.GetState())
+		return controller.Result{}, nil
+	}
+
+	if ok, err := r.reconcileConfiguration(ctx, config); err != nil {
+		log.Warnf("Failed to reconcile configuration: %s, %s", configurationID, err)
+		return controller.Result{}, err
+	} else if ok {
+		return controller.Result{}, nil
+	}
+
+	return controller.Result{}, nil
+
+}
+
+func (r *Reconciler) reconcileConfiguration(ctx context.Context, config *configapi.Configuration) (bool, error) {
+	conn, err := r.conns.Get(ctx, topoapi.ID(config.TargetID))
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return false, err
+		}
+		return false, nil
+	}
+
+	rootPath := &gpb.Path{Elem: make([]*gpb.PathElem, 0)}
+	getRootReq := &gpb.GetRequest{
+		Path:     []*gpb.Path{rootPath},
+		Encoding: gpb.Encoding_JSON_IETF,
+	}
+	root, err := conn.Get(ctx, getRootReq)
+	if err != nil {
+		return false, err
+	}
+
+	if len(root.Notification) == 0 {
+		return false, errors.NewInvalid("notification list is empty")
+	}
+
+	var currentConfigValues []*configapi.PathValue
+	for _, notification := range root.Notification {
+		for _, update := range notification.Update {
+			configValues, err := jsonvalues.DecomposeJSONWithPaths("", update.GetVal().GetJsonIetfVal(), nil, nil)
```

Comment:
This utility function with a different name will be called  using the plugin interface